relay-server/src/metrics_extraction/transactions.rs

```diff
@@ -57,6 +84,96 @@ fn extract_dist(transaction: &Event) -> Option<String> {
     dist
 }
 
+/// Satisfaction value used for Apdex and User Misery
+/// https://docs.sentry.io/product/performance/metrics/#apdex
+#[cfg(feature = "processing")]
+enum UserSatisfaction {
+    Satisfied,
+    Tolerated,
+    Frustrated,
+}
+
+#[cfg(feature = "processing")]
+impl fmt::Display for UserSatisfaction {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            UserSatisfaction::Satisfied => write!(f, "satisfied"),
+            UserSatisfaction::Tolerated => write!(f, "tolerated"),
+            UserSatisfaction::Frustrated => write!(f, "frustrated"),
+        }
+    }
+}
+
+#[cfg(feature = "processing")]
+impl UserSatisfaction {
+    /// The frustration threshold is always four times the threshold
+    /// (see https://docs.sentry.io/product/performance/metrics/#apdex)
+    const FRUSTRATION_FACTOR: f64 = 4.0;
+
+    fn from_value(value: f64, threshold: f64) -> Self {
+        if value < threshold {
+            Self::Satisfied
+        } else if value < Self::FRUSTRATION_FACTOR * threshold {
+            Self::Tolerated
+        } else {
+            Self::Frustrated
+        }
+    }
+}
+
+/// Get duration from timestamp and start_timestamp.
+#[cfg(feature = "processing")]
+fn get_duration_millis(transaction: &Event) -> Option<f64> {
+    let start = transaction.start_timestamp.value();
+    let end = transaction.timestamp.value();
+    match (start, end) {
+        (Some(start), Some(end)) => {
+            let start = start.timestamp_millis();
+            let end = end.timestamp_millis();
+            Some(end.saturating_sub(start) as f64)
+        }
+        _ => None,
```

Comment:
Open question: Should a transaction without duration count towards satisfied users?

Reply:
We do not have to choose, we can simply leave the metric untagged. So I will leave `None` here. This is an exceptional case anyway, because both `timestamp` and `start_timestamp` are required fields: https://develop.sentry.dev/sdk/event-payloads/transaction/.